python-package/lightgbm/dask.py

```diff
@@ -434,25 +434,49 @@ def _predict(
 
 class _DaskLGBMModel:
 
+    # self._client is set in the constructor of classes that use this mixin
+    _client: Optional[Client] = None
+
+    @property
+    def client_(self) -> Client:
+        """Dask client.
+
+        This property can be passed in the constructor or updated
+        with ``model.set_params(client=client)``.
+        """
+        if self._client is None:
```

Comment:
Oh I see, ok. Why should the check be on `._n_features`? That seems kind of indirect. Shouldn't it be on `.fitted_` directly for a check of whether or not the model has been fit?

I'm also confused how you would like `_fit()` to work. I'm currently passing `client=self.client_` into `self._fit()`, and relying on that to resolve whether there is a client stored on the object or `get_default_client()` should be used. This suggested change would break that behavior, because of course the model is probably not fitted yet at the time you call `.fit()`. What do you suggest I do for that?

> BTW, I think it will be great to setup sklearn integration tests at our CI for Dask classes. It will not allow us to be sure that our classes fully compatible with sklearn but at least will check basic compatibility. WDYT

Sure, I think that's fine. I think it's an ok fit for this PR and will add it here.

Reply:
> Why should the check be on `._n_features`?

For historical reasons. `fitted_` was introduced quite recently in our sklearn wrapper. But now you can use it, indeed 👍 .

> I'm also confused how you would like `_fit()` to work.

Oh I see now! Can you pass `self.client` into `_fit()` and there assign result of a check to `self._client`? While `client_` will always return `self._client` or raise error? Just like all other properties in sklearn wrapper.

Reply:
> I think it's an ok fit for this PR and will add it here.

Maybe in a follow-up PR? I'm afraid that it can fail now and stop merging `client` argument migration.

Reply:
I want to be really sure that there is exactly one place where resolve which client to use (https://github.com/microsoft/LightGBM/pull/3883#issuecomment-770554762).

So I don't want to put this code into the body of `_fit()`.

```python
if self.client is None:
    client = default_client()
else:
    client = self.client
```

I think it would work to pull that out into a small function like this:

```python
def _choose_client(client: Optional[Client]):
    if self.client is None:
        return default_client()
    else:
        return self.client
```

Then use that in both the `client_` property (after the fitted check) and in `_fit()`. That would give us confidence that accessing `.client_` returns the same client that will be used in training.

Reply:
OK, sounds good!

Reply:
Added https://github.com/microsoft/LightGBM/issues/3894 to capture the task of adding a scikit-learn compatibility test. I made this a "good first issue" because I think it could be done without deep knowledge of LightGBM, but I also think it's ok for you or me to pick up in the near future (we don't need to reserve it for new contributors).

Reply:
Ok I think I've addressed these suggestions in https://github.com/microsoft/LightGBM/pull/3883/commits/555a57a47345e6b5b1a205b88a4c794bbf096e50

* added an `LGBMNotFittedError` if accessing `.client_` for a not-yet-fitted model
* added an internal function `_get_dask_cluster()` to hold the logic of using `default_client()` if `client` is None
* changed the unit tests to use two different clusters and clients

Note that the change to have two clusters active in the same test will result in a lot of this warning in the logs:

> test_dask.py::test_model_and_local_version_are_picklable_whether_or_not_client_set_explicitly[False-ranking-cloudpickle]
  /home/jlamb/miniconda3/lib/python3.7/site-packages/distributed/node.py:155: UserWarning: Port 8787 is already in use.
  Perhaps you already have a cluster running?
  Hosting the HTTP server on port 40743 instead
    http_address["port"], self.http_server.port

This is not something we have to worry about, and I'd rather leave it alone and let Dask do the right thing (picking a random port for the scheduler when the default one is unavailable) than add more complexity to the tests by adding our own logic to set the scheduler port.